### PR TITLE
Introduced FrontendApplicationStateService for proper handling of the application startup

### LINKED ISSUE #1657: Client "offline" mode falsely reported
I have recently noticed that my client reports being "offline" when it's actually not. It's not just kicking-in momentarily, but rather seems to stay in that mode indefinitely, once it triggers. 

Looking at the backend logs, it seems that the "alive request" is responded-to as expected: 
```
{"name":"Theia","logger":"root","hostname":"elxa610sxy1","pid":16374,"level":10,"msg":"Checking backend connection status. Scheduled an alive request with 1000 ms timeout. []","time":"2018-04-05T09:18:47.944Z","v":0}
{"name":"Theia","logger":"root","hostname":"elxa610sxy1","pid":16374,"level":10,"msg":"Connected to the backend. []","time":"2018-04-05T09:18:47.946Z","v":0}

```

### PR DESCRIPTION
**Pull request by spoenemann**
Fixes #1657. I also removed the `loading` property of ApplicationShell by using the new state handler instead.

### CODE REVIEW

**Comment by spoenemann** [comment]
Also fixes #1749.

File: packages/core/src/browser/frontend-application-state.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+export type FrontendApplicationState =
+    'init'
+    | 'started_contributions'
+    | 'attached_shell'
+    | 'initialized_layout'
+    | 'ready'
+    | 'closing_window';
+
+@injectable()
+export class FrontendApplicationStateHandler {
```
Comment (svenefftinge) [inline packages/core/src/browser/frontend-application-state.ts]: I think `Handler`is not the right term, here. Something that registers for statechange events would be called a handler. Something like 'service' or 'manager' would be more in line with how we name these kind of things.

Comment (svenefftinge) [reply packages/core/src/browser/frontend-application-state.ts]: Or Provider

File: packages/core/src/browser/status-bar/status-bar.ts [hunk 2]
```diff
@@ -70,36 +70,35 @@ export class StatusBarImpl extends VirtualWidget implements StatusBar {
...
         super();
         delete this.scrollOptions;
         this.id = 'theia-statusBar';
     }
 
-    onAfterAttach(msg: Message): void {
-        super.onAfterAttach(msg);
-        this.attached.resolve();
+    protected get ready(): Promise<void> {
+        return this.applicationStateService.reachedAnyState('initialized_layout', 'ready');
```
Comment (simark) [inline packages/core/src/browser/status-bar/status-bar.ts]: Why is it necesary to wait for either `'initialized_layout'` or `'ready'`?  Wouldn't it be necessary to just wait for `'initialized_layout'`?

Comment (spoenemann) [reply packages/core/src/browser/status-bar/status-bar.ts]: This method is also called when the status bar is changed while the application is running. In that case the state is `ready`.

**Pull request merged**